### ISSUE TITLE
"destruct" cannot use injection patterns top-level (but it can nested)

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#13205]
#### Description of the problem

The following code works:
```coq
Lemma foo (a b : nat) : (True /\ Some a = Some b) -> a = b.
Proof.
intros HX.
destruct HX as [_ [= ->]].
```
But this does not:
```coq
Lemma foo (a b : nat) : Some a = Some b -> a = b.
Proof.
intros HX.
destruct HX as [= ->].
```
That is quite surprising -- `destruct` accepts the full grammar of intro patterns, but only nested; top-level it cannot be a `=` pattern.

Only when asking about this on Zulip did I learn about `injection HX as [= ->]`, which indeed works. But this means that even after one learned about the various kinds of intro patterns that let one take apart values in lots of powerful ways, one *still* has to also learn about `injection`, which seems entirely redundant at this point.

This is quite surprising IMO -- for many years I used Coq and the equivalence of `destruct HX as pat` and `revert HX; intros pat` held up. But here it breaks down. The reason seems to be mostly a historic accident to me, namely that there are two top-level entry points `destruct` and `injection`. But I do not think there is any good reason to keep what looks like an artificial restriction of the kinds of patterns that `destruct` can work with. `destruct` should just be "re-branded" as "the tactic to apply an intro pattern to a term", which is already how most Coq users will think about it (it certainly is how I was taught that tactic).

If there is strong resistance to keeping the current behavior, what about at least improving the error message to tell the user to use `injection` instead?

#### Coq Version

Coq 8.12.0

**Comment by herbelin** [comment]
It seems to me that the misunderstanding is on the word `destruct` which is commonly confusing. Technically it is like a `case` (and you would not expect `case HX as [= ->]` to succeed I guess), but in the intuition that it conveys, it is expected to accept any "destruction" rule.

Extending the role of `destruct` could be done of course so that it fits better the intuition. The only worry I had was what word to use then for `case HX as [ ... ]`, so as to keep a parallel with `injection`, `apply in`, `rewrite`

**Comment by JasonGross** [comment]
> and you would not expect `case HX as [= ->]` to succeed I guess

Why not?  (I'm not actually very familiar with into patterns)

**Comment by RalfJung** [comment]
Given that we have this very general language of intro patterns, I would expect every place that I can use intro patterns to do a best effort to support all possible intro patterns. That is more uniform and easier to learn that having to remember which subset of intro pattern syntax is supported where.

So, in `<tactic> as <pattern>`, I think the `<tactic>` should have zero influence on what `<pattern>`s are supported.

`apply in` and `rewrite` are syntactically very different from `destruct` and `injection`, so I do not see the parallel here. `destruct` and `injection` I would expect to only differ in there non-`as` form. Indeed until this morning, I did not even know that `injection` *has* an `as` form. I never saw the need for that, after all I can just use intro patterns instead...

**Comment by herbelin** [comment]
Ralf, if `destruct H as pat` is the same for you as `injection H as pat`, what role do you leave to the word `destruct` and `injection`?

If by compositional you mean to see `H as pat` as a standalone tactics, see this coqdev [discussion](https://sympa.inria.fr/sympa/arc/coqdev/2016-02/msg00009.html) from 2016 which I believe proposes what you are asking for (with notation `on H do pat`).

**Comment by herbelin** [comment]
Well, I made a prototype for such standalone `on H do pat` at #13206.

**Comment by JasonGross** [comment]
> what role do you leave to the words destruct and injection?

`destruct x` is different from `injection x`, right?

**Comment by herbelin** [comment]
> `destruct x` is different from `injection x`, right?

Right.

> Or do you mean, what's the point of having two variants involving patterns that so the same thing?

Yes.

**Comment by RalfJung** [comment]
> >  Or do you mean, what's the point of having two variants involving patterns that so the same thing?
>
> Yes.

Indeed I don't see any point. `destruct X as pat` can already express everything here. Which is why I didn't expect `injection X as pat` to exist. I assume it exists mostly as a historical accident from before intro patterns were able to perform injection.

> If by compositional you mean to see H as pat as a standalone tactic, see this coqdev discussion from 2016 which I believe proposes what you are asking for (with notation on H do pat).

I mean, of course we could solve the existing non-compositionality of these tactics by adding a third tactic, re-writing all teaching material, waiting a year, and adjusting all our code to use it. Basically every occurrence of `destruct ... as` should be changed to the new tactic I think. (And `destruct` without `as` should be avoided as it relies on auto-generated names.) That would ensure that people no longer run into such inconsistencies.

But wouldn't it be much quicker to just permit `destruct` to run arbitrary destruct patterns?

Also, with such a tactic, what is even the justification for having `destruct as` and `injection as`? When would it ever be useful to give a destruct pattern but also redundantly encode the information of the topmost action the destruct pattern is performing?

Finally let me note that I feel the same, in principle, about `destruct X as ->` and `destruct X as H%lemma`. Note that the former is less powerful than `rewrite` as it lacks chaining and position specifiers, so this does not make `rewrite` obsolete -- but `destruct X as ->` is the natural consequence of "`destruct` takes a destruct pattern" and "destruct patterns can rewrite", so it is very weird to me that it would not work, just because there is another way of achieving this effect.

To give an example, programming languages typically allow programmers to write `x+0`, even though there is another way of writing the same thing, namely `x`. But "this introduces multiple ways to say the same thing" is not a sufficient argument for introducing non-compositional behavior: `+` should work on any two numbers, and `as` should work with any intro pattern.

**Comment by herbelin** [comment]
I agree that we could have provided `injection as` by generalizing the `as` of `destruct as` rather than by generalizing `injection`. But isn't it before all a question of how we explain things? There are five kinds of "action" patterns (`[ ... ]`/`( ... )`, `[= ...]`, `->`/`<-`, `%` and `_`). They are all five supported by `intros` but each of the tactics `destruct as` and `injection as` (as well as `apply in`, `rewrite`, `clear` if we include degenerate cases with one or zero subgoals) support only the kind of pattern which matches the name of the tactic.

Of course, by saying so, I identify `destruct` with a role of case analysis, which is disputable, at least from the English point of view, since from the English point of view, `destruct` would somehow be meaningful for any of these 5 "destructive" actions. Is it this point that you are questioning? And, if not, would you then expect that `injection as` takes any pattern too?

**Comment by RalfJung** [comment]
> They are all five supported by intros but each of the tactics destruct as and injection as (as well as apply in, rewrite, clear if we include degenerate cases with one or zero subgoals) support only the kind of pattern which matches the name of the tactic.

That is not entirely correct; each of the tactics only supports one kind of pattern *at the top level* but supports arbitrary patterns in a nested position.
If the argument is "destruct should only do pattern matching on inductives", then how does it make sense that `destruct` supports injection for *nested* equalities?

> Of course, by saying so, I identify destruct with a role of case analysis, which is disputable, at least from the English point of view, since from the English point of view, destruct would somehow be meaningful for any of these 5 "destructive" actions. Is it this point that you are questioning?

That is also part of the point -- in my head, "intro pattern" and "destruct pattern" were synonyms until this discussion started, since these patterns are all about "somehow taking the value apart".

**Comment by herbelin** [comment]
> That is not entirely correct; each of the tactics only supports one kind of pattern at the top level but supports arbitrary patterns in a nested position.

Right.

I remain reticent to generalize `destruct as` because I don't see why we wouldn't also generalize `injection as`. This is why it seems better to me to provide an `on H do pat` (or maybe `use H as pat`, so as to echo to the existing `as`) plus, as you suggest, an error message to tell the user to use `injection` or `destruct` (or `use as`) when using `destruct` or `injection` with a toplevel pat which does not match the purpose of the tactic.

Would that be acceptable?

**Comment by RalfJung** [comment]
A specific error message is certainly an improvement that we seem to agree on, so it is an obvious first step.

For the new tactic, my comment above still stands -- once that tactics exist, do you see any reason for e.g. teaching material to mention `destruct as` or `injection as` at all, or for idiomatic code to use them?

**Comment by herbelin** [comment]
> A specific error message is certainly an improvement that we seem to agree on, so it is an obvious first step.

I made #13212 with message "Found an injection pattern while a disjunctive/conjunctive pattern was expected; use tactic injection instead.". Don't hesitate to comment there to make improvements.

**Comment by ppedrot** [comment]
> at least from the English point of view

I am not sure what the linguistic considerations about a natural language (which, in addition, has a strong anti-prescriptivist tradition and no central Academy) have anything to do about the syntax of a formal language such as Ltac.

I side with @RalfJung about the fact that restricting the toplevel shape of destruct patterns in an arbitrary limitation, and see the obvious downsides of the introduction of yet another syntax, the **mother of modularity issues**, when we can simply extend the abilities of `destruct` in a fully backwards compatible way.

**Comment by herbelin** [comment]
> restricting the toplevel shape of destruct patterns in an arbitrary limitation

The way I think at it, there is no restriction or arbitrary limitation since `destruct` is for inductive-type "destruction". The way I see it, what you are asking is not to remove a restriction but to add a tolerance for any other (non-inductive-type) "destructive" pattern.

So that I understand the global picture that you have in mind, the simplest would maybe be that you propose a change of the reference manual (or even a stand-alone PR) where `destruct` accepts any destructive pattern? Personally, I would not know how to explain it. E.g., would it apply to `induction` as well, would it apply to n-ary `destruct` and n-ary `induction`, would it apply to `injection` as well? Would you accept `destruct c as _`? Would you accept `destruct c as ?id`?

**Comment by ppedrot** [comment]
Proposal: `destruct ... as ...` is the universal syntax for the set of "elimination rules" defined in terms of intropatterns, which has the behaviour implemented in your draft PR but without any change of syntax.

> Personally, I would not know how to explain it

IIUC, @RalfJung was surprised that it was not working that way after realizing we already tolerate rich intropatterns outside of the toplevel. If anything, this shows that the current behaviour is the weird one.

**Comment by RalfJung** [comment]
> The way I think at it, there is no restriction or arbitrary limitation since destruct is for inductive-type "destruction".

On a purely technical level, the implementation of `destruct` specifically checks for this condition before evaluating the pattern -- and it does not do this check because somehow the pattern would not work, or anything like that. If we removed the check, everything would still work.

In that sense, the limitation feels rather arbitrary.

**Comment by herbelin** [comment]
@RalfJung, @ppedrot: I don't know what to tell more. There is the draft PR #13206 and if you want to use `destruct as` in place of `on do`, feel free to push on top of the PR. As I said, I would not know how to modify the reference manual to explain the story about this generalization of `destruct` in a direction which in my view is orthogonal to all other existing generalizations of `destruct`. So, really, it is your call.

**Comment by herbelin** [comment]
Alternatively, maybe are you saying that we should not care if the basic `destruct c as pat` is extended in orthogonal ways. It is so intuitive to think at `destruct` as a general destruction tactic that it is worth overloading the basic `destruct c as pat` form even though it can only have an "inductive-type-destruction" meaning in its more general forms?

**Comment by RalfJung** [comment]
> It is so intuitive to think at destruct as a general destruction tactic that it is worth overloading the basic destruct c as pat form even though it can only have an "inductive-type-destruction" meaning in its more general forms?

Which more general forms are you referring to here?

**Comment by herbelin** [comment]
What I basically said above, `destruct c as pat, c as pat'`, `destruct c as pat using foo`, `destruct c as pat eqn:H`, `induction c as pat`, `induction c as pat, c' as pat'`, `destruct c as pat in H |- *`, ...

**Comment by RalfJung** [comment]
All of these `pat` could support arbitrary intro patterns without technical trouble, right? So what is the issue?

**Comment by herbelin** [comment]
That's right, one could give a role to `eqn:H` and `in H |- *` at least to `injection` by inverting the proof of injectivity in its dependent occurrences. The ability to do such inversion is actually certainly something nice to have.

For `->` or `apply in`, or even `specialize`, do you have some particular semantics in mind? How would you interpret the dependent occurrences of `c` after the introduction pattern has been applied?

We could also give to `using foo` the general meaning of a prior `refine (foo ... c ...)` when some well-chosen `...` can be found.

On the other side, would you accept `induction c as [=]` and if yes with which meaning? Also, what kind of destruction would do `destruct c` when no pattern is given?

**Comment by RalfJung** [comment]
> For -> or apply in, or even specialize, do you have some particular semantics in mind? How would you interpret the dependent occurrences of c after the introduction pattern has been applied?

I don't think I understand the question... I imagine that there is a way in Coq to say "run this intro pattern on that term", is that not the case? Probably there are many places where Coq accepts intro patterns that I am not even aware of. For example, where is there a pattern in `apply in`?

> On the other side, would you accept induction c as [=] and if yes with which meaning? 

I mean, that is certainly weird, but is it a problem? It would be equivalent to something like `induction c as X; [injection X as [=] ]`.

> Also, what kind of destruction would do destruct c when no pattern is given?

That is really the odd one out, isn't it? `destruct` without a pattern? FWIW in our developments we tell people to not do this, as it leads to unpredictable variable names.

**Comment by herbelin** [comment]
I was thinking e.g. to a goal of the form, where the destructed object is dependent:
```
Theorem foo (x:nat) : x = x.
Proof.
destruct x as []%S in |- *. (* standing for a dependent form of "apply S in x" *)
```
To be able to give meaning to the dependency of `x` in the conclusion (and to an `in |- *` clause), one would need a way to invert `S` so that `x` is rephrased in term of the action of the pattern, i.e. in terms of `S x` (which for `S` is possible, leading to `x : nat |- pred x = pred x`, but which is not generally possible).

> That is really the odd one out, isn't it? destruct without a pattern? FWIW in our developments we tell people to not do this, as it leads to unpredictable variable names.

Not telling that we should rely on unpredictable name but there are `2600` such `destruct` in the Coq `theories` directory (about which I'm sure many are name-"safe", Jasper has maybe statistics).

This being said, on the observation that many people find intuitive that `destruct c as pat` accepts any pattern, I'd be ok to rally the majority. This could be explained as a special case in the documentation.

**Comment by JasonGross** [comment]
> I was thinking e.g. to a goal of the form, where the destructed object is dependent:
> 
> ```
> Theorem foo (x:nat) : x = x.
> Proof.
> destruct x as []%S in |- *. (* standing for a dependent form of "apply S in x" *)
> ```
> 
> To be able to give meaning to the dependency of `x` in the conclusion (and to an `in |- *` clause), one would need a way to invert `S` so that `x` is rephrased in term of the action of the pattern, i.e. in terms of `S x` (which for `S` is possible, leading to `x : nat |- pred x = pred x`, but which is not generally possible).

As a total tangent, I think it would be really cool if intropatterns here could be extended to support this case using something like typeclasses.  Here's a sketch of how it might work:

```coq
Class Sect {A B : Type} (s : A -> B) (r : B -> A) :=
  is_sect : forall x : A, r (s x) = x.

Global Instance S_is_sect : Sect S pred := fun x => eq_refl.

Lemma sect_ind {A B : Type} (s : A -> B) {r : B -> A} {H : Sect s r}
      (P : A -> Type)
  : (forall x, P (r x)) -> (forall x, P x).
Proof.
  intros p x; specialize (H x); specialize (p (s x)); case H; exact p.
Defined.

Ltac dependent_apply_in H x := induction x using (@sect_ind _ _ H _ _).

Goal forall x : nat, x = x.
  intro x.
  dependent_apply_in S x.
```

**Comment by RalfJung** [comment]
> To be able to give meaning to the dependency of x in the conclusion (and to an in |- * clause), one would need a way to invert S so that x is rephrased in term of the action of the pattern, i.e. in terms of S x (which for S is possible, leading to x : nat |- pred x = pred x, but which is not generally possible).

I'm afraid this is way over my head. I have never seen `in |- *` clauses before and I do not know what you mean by "a dependent form of "apply S in x"". I looked up `in` and it seems to control which parts of the goal are adjusted by the case distinction; I do not understand how that would have interesting interactions with top-level-injection patterns.

Your code example should behave the same as
```
Theorem foo (x:unit * nat) : x = x.
Proof.
destruct x as [_ []%S] in |- *.
```
which is syntactically accepted but then fails saying
```
Cannot change n, it is used in conclusion.
```

**Comment by herbelin** [comment]
> `Cannot change n, it is used in conclusion.`

Well, I'm getting short in arguments...

>  I do not understand how that would have interesting interactions with top-level-injection patterns.

I don't know if injection on equalities dependent in a goal is per se interesting, and in particular whether it is interesting to select only a few occurrences rather than all (since we would loose reversibility by keeping only a selection), but dependent injectable equalities can certainly show up sometimes (e.g. in a HoTT context).

> As a total tangent, I think it would be really cool if intropatterns here could be extended to support this case using something like typeclasses. Here's a sketch of how it might work:

Cool indeed. Not clear how to proceed. Should we have a standard class `Sect` (like we have `Reflexive`)? Should we have an automatic recognition of injective constructor with automatic registration in this class? Should we re-implement apply-in in Ltac2? Or hard-wire the class `Sect` in the OCaml code for apply-in?

**Comment by JasonGross** [comment]
> Should we have a standard class Sect (like we have Reflexive)?

If we want to have this extension, probably, yes.

> Should we have an automatic recognition of injective constructor with automatic registration in this class? 

I'd like to see a more general mechanism here where users can register arbitrary constants, hints, and instances, to be solved by arbitrary tactics, to run on each inductive declaration, conditional on some boolean setting.  Then things like `Set Boolean Equality Schemes` and `Set Decidable Equality Schemes` can be moved to user-space.

> Should we re-implement apply-in in Ltac2?

Perhaps, but only if it doesn't incur overhead.  In general I think it would be good to move as much of the tactic implementation code as possible to Ltac2, so that it's more easily accessible to users.

> Or hard-wire the class Sect in the OCaml code for apply-in?

Is we do this it should be via `Register`, I think.

**Comment by herbelin** [comment]
Is there a conclusion to the original issue?

I was proposing a new generic syntax (`on c do pat`, or `use c as pat`), coming a priori with no `eqn:`, `in`, etc. variants, but:
- it is pointed out that this new syntax could actually also support the same variants as `destruct` does (even if it fails in some cases, but not more than when using nested patterns in `destruct`)
- the alternative proposal is then to save a new word and reuse the appealing name `destruct` instead with the convention that `destruct c` alone defaults to `destruct c as []`.

We would then have:
- `destruct c as [...]` as itself
- `destruct c as [= ...]` = `injection c as [= ...]`
- `destruct c as ->` = `rewrite c` (for `c` of quantifier-free type) or `subst x` if `c` has a type of the form `x = t`
- `destruct H as _` = `clear H`
- `destruct H as H'` = `rename H into H'`
- `destruct H as H%foo` = `apply foo in H` (with a proposal to support the case when `H` is dependent and `foo` injective)
- and their variants, when applicable

Is that a correct conclusion?

**Comment by Zimmi48** [comment]
From this discussion, I see three possible conclusions, all with their own appeal:

- Interpret `destruct` as *the* general tactic to use "intro patterns" (that we could indeed rename into "destruct patterns").

- Tactics supporting a restricted set of intro patterns should parse all intro patterns and then produce a useful error message when the wrong intro-pattern is used. (This conclusion could be combined or not with the previous one.)

- Tactics supporting the `as` clause should all be lax (all behave like the hypothesized generalized `destruct`) but raise a warning when they are used with the "wrong" top-level intro-pattern. E.g. `destruct ... as [=]` would work but would raise a warning suggesting to use `injection ... as [=]` instead. And `injection ... as ->` would work but raise a warning suggesting to use `rewrite ... as ->` instead (the `as` clause is not supported with `rewrite` as of today though).

A few things to note (and possibly to fix first) though:

- the destruct-intro-pattern `[ ... | ... | ... ]` used with `induction` currently has a different meaning than when used with any other tactic (in particular when used nested);
- the `inversion` tactic does not correspond to any intro pattern yet, but nevertheless already supports the `as` clause;
- the `injection`  tactic supports both the `[=]` intro pattern and a legacy syntax *à la* inversion, without raising any warning. #9288, who introduced support for the `[=]` intro pattern in this `as` clause, initially proposed a deprecation of the legacy syntax, but this was eventually backtracked on. Already during this discussion, the redundancy of the tactic name + the top-level intro-pattern was a point of the discussion.

**Comment by RalfJung** [comment]
@herbelin that reflects my original proposal, albeit with way more detail than I could have supplied. :)

**Comment by herbelin** [comment]
See also #7885.

**Comment by RalfJung** [comment]
Good catch... looks like I had the same problem 2 years ago and forgot about it in the mean time.^^

**Comment by herbelin** [comment]
> and forgot about it in the mean time

and so did I...

**Comment by Zimmi48** [comment]
The suggestion of @RalfJung was then the same two years ago and it was supported at that time by @anton-trunov.

At that time, @herbelin was already tempted to generalize the meaning of `destruct`, but expressed the issue that we would lose the current correspondence between intro-patterns and tactics. As I understand it, this is still the same issue that prevents him from being fully convinced by this change.

Note that #13212 is a step in the direction of implementing the second-bullet point from https://github.com/coq/coq/issues/13205#issuecomment-711896908. More generally, if we decide not to make the proof language more lax, we can seriously improve the user experience by providing guidance in the error messages by relying on common mistakes to guess what the user really wants.

**Comment by herbelin** [comment]
> if we decide not to make the proof language more lax

As author of  #13305, I remain perplex: if ever there ought to be a tactic accepting all introduction patterns, would it be better to extend/overload `destruct` or to have this specific `on do`/`use as` variant of #13305. I may be wrong but it seems that the debate is both about how important it is to limit the number of tactic names and about how intuitive a tactic name should be.

**Comment by RalfJung** [comment]
I think backwards compatibility also factors in. The reason I think `destruct` should be universal in terms of patterns is that it is already the most widely used way to execute a pattern on a term -- so this requires the least amount of adjustment of existing intuition, code, and teaching material.

**Comment by herbelin** [comment]
So, you want to reverse the point of view, and explain that if `destruct H` (without `as`) defaults to case analysis, it is a choice which is not related to the name "destruct" but an arbitrary choice, just because it is convenient to have a defective case?

I guess that you then want also to clearly distinguish `destruct` from `induction`? (Note that they are implemented the same in the code, and this incidentally certainly induces a bias in my way of seeing things.)

**Comment by RalfJung** [comment]
> So, you want to reverse the point of view, and explain that if destruct H (without as) defaults to case analysis, it is a choice which is not related to the name "destruct" but an arbitrary choice, just because it is convenient to have a defective case?

I want to to say that it is a historic accident, essentially.

I wouldn't make this proposal in a clean-slate design, but we are not doing clean-slate design here. (In a clean-slate design, I'd vote for requiring `as`. I see little point of supporting `destruct` without `as`. But this is a futile part of the discussion since we are bound by backwards compatibility.)

**Comment by Zimmi48** [comment]
@herbelin `#13305` is probably not the issue you intended to refer to.

@RalfJung Despite being quite OK myself with the generalization of the meaning of `destruct`, I don't buy *at all* the backward compatibility argument. Nothing prevents introducing a new tactic now. Anyone using it would need Coq 8.14 or later, that's all. Existing scripts would continue to work as is since we are not talking about removing the `destruct` tactic. On the contrary, if we were making the `destruct` meaning more general, it would be less clear at first sight that a script requires Coq 8.14 or later because it uses a "destruct pattern" that used to not be supported.

**Comment by ppedrot** [comment]
>  I don't buy at all the backward compatibility argument.

Repeat after me: *syntax is not modular*. Every time you add a new tactic defined with a primitive parsing rule, some code might break out there.

**Comment by RalfJung** [comment]
Backward compatibility is not just a matter of keeping existing scripts working, though. If such a unified tactic exists, IMO there is no reason to ever use `destruct ... as`, which means it should not be used an existing code should be updated to remain consistent with the latest recommendations. If Coq 8.14 made this change, then as soon as Iris dropped support for 8.13 I'd push for changing all `destruct` (and the few `injection`) to the new tactic, because the alternative is to have a messy mixture of old code and new code -- and, consequently, new users still learning the old syntax.

The amount of ecosystem churn created by porting all `destruct ... as` to the new tactic is enormous. The cost of *not* doing this to have to forever explain to new users the history of why there are 18 ways to do the same thing, and which of them they should use.

**Comment by ppedrot** [comment]
Side comment as I am rereading the whole discussion, and I thought it was mentioned but can't find it back, so let me say it out loud again.

The culprit of this messy confusion is `injection`, not `destruct`. It should **never** had been the case that `injection foo as [= pat]` is equivalent to `injection foo as pat`. The semantics of injection should be injection as a toplevel pattern + whatever follows, that is `injection foo as pat` **ought to be equivalent to** `destruct foo as [= pat]` by the very definition of `injection`. But now we have these Frankenstein tactics, so it's too late to fix.

**Comment by RalfJung** [comment]
> injection foo as [= pat] is equivalent to injection foo as pat

Wait, these two are equivalent? I had no idea.^^ I guess that explains why some people though `destruct` would add an implicit `[...]` around the given pattern. Yeah some of the comments I got make much more sense now.

**Comment by Zimmi48** [comment]
> The amount of ecosystem churn created by porting all `destruct ... as` to the new tactic is enormous. The cost of _not_ doing this is to have to forever explain to new users the history of why there are multiple ways to do the same thing, and which of them they should use.

OK thanks for the explanation (which I do buy). I think backward compatibility was not the right term to use then to express this idea.

**Comment by ppedrot** [comment]
@RalfJung for reference, this is https://github.com/coq/coq/pull/9288.

**Comment by Zimmi48** [comment]
>for reference, this is #9288.

which I've already referenced in https://github.com/coq/coq/issues/13205#issuecomment-711896908...

Did you really re-read the entire discussion @ppedrot?

**Comment by ppedrot** [comment]
My bad, but this discussion is starting to be big enough to write a novel, and it's spread across various PRs. Not easy to spot the relevant information. (And @RalfJung's reaction seems to indicate that I am not the only one to get lost.)

**Comment by herbelin** [comment]
@RalfJung 
> I want to to say that it is a historic accident, essentially.

Historic yes but accident no. `destruct H` predates `destruct H as`.

The idea that `destruct as` is more primitive is somehow a mental reconstruction coming from the practice of users.

> I see little point of supporting `destruct` without `as`

Do you see a point in supporting `injection` without `as`? If yes, then it is that you really think at `destruct` the way as I think to `do on`.

@Zimmi48:
> @herbelin #13305 is probably not the issue you intended to refer to.

Github tries to be too smart and change the PR numbers alone (and I fail to see it). I meant #13206, sorry (fixed).

@RalfJung 
> The amount of ecosystem churn created by porting all destruct ... as to the new tactic is enormous. 

Why to port all `destruct`? And if so, why not porting all `apply in`, all `injection`, all relevant `subst`, which would be derivable too? All these names bear an intuition which I believe is good for pedagogy. Why not to tell that there are different levels of language. One with words, one with symbolic notations, and `on do` is the keyword allowing to use the symbolic notation matching the corresponding words.

@ppedrot
>  that is `injection foo as pat` ought to be equivalent to `destruct foo as [= pat]`

I'm lost. Then, to follow this analogy, what should have been the syntax without `[ ... | ... ]` for `destruct foo as [ ... | ... ]`?

More generally, if we want to understand each other, we have to consider that there are (at least) two models in play (`destruct` as the swiss knife and things like `injection` are redundant vs `destruct` as case analysis and `on do` as the swiss knife). The question then, is to evaluate the pros and cons of each model.

**Comment by Zimmi48** [comment]
The only important point about PMP's remark (IMHO) is what I already said previously: there still is an issue with some tactics supporting a "legacy" `as` clause that doesn't correspond to the model where all the tactics are encompassed in symbolic notations. This is the case for `inversion` (which has no corresponding pattern), `induction` (which piggy-backs on the and/or patterns of `destruct`) and `injection` (which has its own intro pattern but also supports a legacy syntax in the `as` clause both for compatibility and possibly to simplify the use of the `as` clause for beginners---although good error messages could be a solution too).

**Comment by herbelin** [comment]
For `induction`, my idea was to use `[ ... | ... ]_ind` seeing `_ind` as a subscript (but I'm open to many other possible ideas). For `inversion`/`inversion_clear` (and `dependent destruction`), my view is that what they provide should be provided by the case analysis tactic (i.e. `destruct` in the current model) without needing a different concept (but the implementation of this old "small-inversion"-based idea is unfortunately still in progress).

@zimmi48: do you foresee a possible concrete outcome to this discussion?

**Comment by RalfJung** [comment]
> Do you see a point in supporting injection without as?

No strong opinion. This is probably more likely to be useful without `as` than `destruct`, because it only produces things whose type is in `Prop`, which have a higher chance of being useful without being named. Right now it is also more "magic", running injectivity recursively, but I think that's actually a bad call (orthogonal to our discussion here though).

> Why to port all destruct?

To have a consistent coding style. To be able to teach new people that start to work with or on Iris something sensible. For better or worse, people learn by looking at existing code and copying it, and we want them to learn the right thing.

> And if so, why not porting all apply in, all injection, all relevant subst, which would be derivable too? All these names bear an intuition which I believe is good for pedagogy.

The only intuition required here is "run an intro pattern on a term". Having to develop separate intuition for `destruct` and `injection` is mental baggage that only makes Coq harder to learn.

`subst` is very different as one does not name the hypothesis being used (`subst x` searches for `x = _`). For `apply in`, the key is not having to repeat the hypothesis name.

> induction (which piggy-backs on the and/or patterns of destruct)

Woah, wait, induction? Induction is way more than just an intro pattern. I would not expect that to be unified into the same framework, this seems really odd. For example, with induction one expects good support for appropriately generalizing the goal before applying the induction principle. `induction ... using` is also a key feature, and showcases the tactic's ability to infer the inductive predicate.

There's an intro pattern as part of it, and that should of course support all kinds of intro patterns, but I don't think there should be an intro pattern that does induction.

**Comment by Zimmi48** [comment]
> I don't think there should be an intro pattern that does induction

OK, it's more the inconsistency of not having an intro-pattern for `induction` and `inversion` but still having an `as` clause for these tactics which I feel problematic.

BTW, I've never liked the `as` clause of `induction` because it is not natural to introduce the variables and the induction hypothesis for a given case in a single list (where the induction hypothesis is not distinguished). It's a bit difficult to anticipate what you need to write (like with `inversion` but less terrible). Therefore, I would not support introducing a `[ ... | ... ]_ind` pattern. IMHO this pattern (if it should exist) should be made more intuitive.

**Comment by Zimmi48** [comment]
>do you foresee a possible concrete outcome to this discussion?

I think we all want a generic tactic allowing to apply any pattern to a hypothesis. As mentioned previously, this tactic actually already exists, it is `specialize` but this name sounds inappropriate to advertise it as *the* generic tactic. The debate is therefore between generalizing `destruct`, because its name sounds rather good for this use case, and introducing a new tactic (`on ... do ...` or `use ... as ...`, the latter having my preference because it keeps the `as` clause). IMHO we need more opinions. Therefore, a strategy could be to summarize the main points in favor of each option into a CEP, then to advertise this CEP widely to get a maximum of feedback from users.

In addition / in parallel, I think it would be a good idea to make the `as` clause universal in terms of *parsing*. I.e. the `as` clause would always accept every pattern and each tactic would error (with a useful error message) on the patterns that it doesn't support.
